### DISCUSSION TITLE
RequiredEntityData types required property values as null

### DISCUSSION THREAD

**Discussion by snigdha920** [mikro-orm/mikro-orm#3290] [Q&A] [answered]
**Describe the bug**

I've been migrating from MikroORM v4 to v5. 
When I type parameters using the `RequiredEntityData` type, it types the required properties as `sometype | null`.
I dug around a bit, and it turns out this happens because in [this](https://github.com/mikro-orm/mikro-orm/blob/424e0bb5c50248ac2d45da55dd1a451663944d31/packages/core/src/typings.ts#L186) line, the required properties have been given the type of [`EntityDataItem<T[K]>`](https://github.com/mikro-orm/mikro-orm/blob/424e0bb5c50248ac2d45da55dd1a451663944d31/packages/core/src/typings.ts#L168) which has a `null` in its type. 

This becomes an issue especially when I have to pass such parameters to a method (eg. a constructor) that does not accept `null` for the required properties and the parameters are typed as `RequiredEntityData`.

**To Reproduce**
1. Type an object using `RequiredEntityData`
2. You will see that the required properties get typed as `sometype | null`

**Expected behavior**

The required properties should not be typed as `null`.

**Additional context**

I checked the issues, but I didn't see that this had been raised before. This seems like an issue that would have come up sooner, so I'm not sure if I'm going about it correctly.

If you agree that this is in fact an issue, I'd be happy to open a PR for this. I think just changing 

```
export type RequiredEntityData<T> = EntityData<T> & { [K in keyof T as RequiredKeys<T, K>]: EntityDataItem<T[K]> };
```

to: 

```
export type RequiredEntityData<T> = EntityData<T> & { [K in keyof T as RequiredKeys<T, K>]: Omit<EntityDataItem<T[K]>, null> };
```

should do the trick. 

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.7.4 |
| mikro-orm | 5.2.2 |
| your-driver | 5.2.2 |

**Comment by B4nan** [comment, marked as answer]
Can you share some code? At least entity definition and a snippet of what you are exactly doing.

**Reply by snigdha920** [reply]
Yes, apologies for not doing that already. My
- entity definition:

```
@Entity({ collection: 'users', customRepository: () => UserRepository })
export class User extends AbstractEntity<UserEntityOptionalProps> {
  @Property()
  @IsString()
  @Unique()
  public sub!: string;

  // more fields

  constructor(
    sub: string,
  ) {
    super();
    this.sub = sub;
  }
}
```

- I define a User object in a test and type it like:

```
export const adminUserData: RequiredEntityData<User> = {
  sub: '123456789',
  // more fields
};
```
- Lastly, when I pass this object to the constructor like so, I get the error: 
```
export const adminUser = new User(
  adminUserData.sub, --> Error: Argument of type 'string | null' is not assignable to parameter of type 'string'. Type 'null' is not assignable to type 'string'
  // more fields
);
```

If I go into the types [here](https://github.com/mikro-orm/mikro-orm/blob/424e0bb5c50248ac2d45da55dd1a451663944d31/packages/core/src/typings.ts#L186), and change the definition of `RequiredEntityData` as I mentioned in the additional context above, it fixes the issue.

**Reply by B4nan** [reply]
Should be addressed by https://github.com/mikro-orm/mikro-orm/commit/e7843fbe2f02ee8922d99a2f0209e4e261898a66

**Reply by snigdha920** [reply]
I haven’t tried it yet, but I think it should work. Thank you for your prompt response!

**Reply by snigdha920** [reply]
Hi again Martin, I noticed that you haven't released a newer version of the package after this commit was merged: 

<img width="2032" alt="image" src="https://user-images.githubusercontent.com/62167899/178579935-605a19b1-5dfb-4e4c-8b58-21fed76facec.png">

I spent a few hours trying to figure out why the v5.2.3 wasn't being installed properly - after which I realised that it was being installed properly, it just didn't include this fix I wanted. 🥲

I request you to please release a newer version as and when you have time. Thank you for your support!
